Mathlib/NumberTheory/DirichletCharacter/Basic.lean

```diff
@@ -74,4 +74,47 @@ lemma changeLevel_trans {m d : ℕ} (hm : n ∣ m) (hd : m ∣ d) :
     changeLevel (dvd_trans hm hd) χ = changeLevel hd (changeLevel hm χ) := by
   simp [changeLevel_def, MonoidHom.comp_assoc, ZMod.unitsMap_comp]
 
+lemma unitsMap_comp {d m : ℕ} (hm : n ∣ m) (hd : m ∣ d) :
+    (ZMod.unitsMap hm).comp (ZMod.unitsMap hd) = ZMod.unitsMap (dvd_trans hm hd) := by
+  simp only [ZMod.unitsMap_def]
+  rw [← Units.map_comp]
+  exact congr_arg Units.map <| congr_arg RingHom.toMonoidHom <| ZMod.castHom_comp hm hd
+
+lemma ofUnitHom_eq {m : ℕ} (hm : n ∣ m) (a : Units (ZMod m)) :
+    (changeLevel hm χ) a = χ a := by
+  simpa [changeLevel_def, Function.comp_apply, MonoidHom.coe_comp] using
+      toUnitHom_eq_char' _ <| ZMod.IsUnit_cast_of_dvd hm a
+
+/-- χ₀ of level d factors through χ of level n if d ∣ n and χ₀ = χ ∘ (ZMod n → ZMod d). -/
+structure FactorsThrough (d : ℕ) : Prop :=
+  /-- `d` divides `n`, so `d` is a potential level through which `χ` factors. -/
+  (dvd : d ∣ n)
+  /-- A character `χ₀` of level `d` such that `χ` is just the change in level of `χ₀` to `n`. -/
+  (indChar : ∃ χ₀ : DirichletCharacter R d, χ = changeLevel dvd χ₀)
```

Comment:
Maybe split this into `χ₀ : DirichletCharacter R d` and `eq_changeLevel : χ = changeLevel dvd χ₀`?
I assume that later on, you will want to use the actual lower-level character and not just its existence; then this would save you deconstructing the existence statement.

Reply:
I tried splitting it like this first:
```lean
/-- χ₀ of level d factors through χ of level n if d ∣ n and χ₀ = χ ∘ (ZMod n → ZMod d). -/
structure FactorsThrough (d : ℕ) : Prop :=
  /-- `d` divides `n`, so `d` is a potential level through which `χ` factors. -/
  (dvd : d ∣ n)
  /-- A character `χ₀` of level `d` ... -/
  (χ₀ : DirichletCharacter R d)
  /-- ... such that `χ` is just the change in level of `χ₀` to `n`. -/
  (eq_changeLevel : χ = changeLevel dvd χ₀)
```
but I got:
```
(kernel) failed to generate projection 'DirichletCharacter.FactorsThrough.χ₀' for 'DirichletCharacter.FactorsThrough', type is an inductive predicate, but field is not a proposition
```
but that didn't work, probably not good to mix data and Proposition.
So I did a slightly different approach, see the current state.

Reply:
solved by the approach below